### ISSUE TITLE
MLX default weight init seems incorrect

### ISSUE THREAD

**Issue by ag487** [ml-explore/mlx#2001]
Seems the default mlx weight initialization is incorrect. Specifically github mlx/python/mlx/nn/layers/linear.py and other similar references.

Generally the starting point for weight init is a normal distribution with mean = zero and variance = 1 / input_dim  or 2 / input_dim (with relu). The issue is discussed in papers like Xavier Glorot "Understanding the difficulty of training deep feedforward neural networks" or Kaiming He "Delving Deep into Rectifiers.."

Like many, mlx uses a uniform distribution to approximate a normal distribution. In this case, mlx uses a uniform distribution with bounds sqrt(1 / input_dim). Mathematically the variance of a uniform distribution with bounds a b is (b - a)^2 / 12. To maintain a zero mean, we set a = -b. To achieve variance = 1 / input_dim = b^2 / 3. Requires bound b = sqrt(3 / input_dim).

As a comparison, look at github pytorch/torch/nn/modules/linear.py the func reset_parameters() uses init.kaiming_uniform_(). Then within pytorch/torch/nn/init.py kaiming_uniform_() multiplies the bound by sqrt(3).

MLX should make a similar adjustment. General training accuracy should improve as a result.

**Comment by angeloskath** [comment]
Indeed our initialization is `sqrt(3)` or someone could even say `sqrt(6)` off from the standard kaiming uniform. To be honest I think at some point it was using normal instead of uniform which made a bit more sense but still.

Personally I don't think this matters as much. Since these analyses both the normalization layers and optimizers have actually changed significantly that I don't think a factor of `~1.5` is going to have a big impact. If people disagree we could fix the default init fairly easily.

**Comment by AvikantSrivastava** [comment]
Hey @awni @angeloskath @ag487 I have raised the PR https://github.com/ml-explore/mlx/pull/2071/files
That adds a parameter for `kaiming_init` 
```
layer = nn.Linear(input_dims=4, output_dims=8, kaiming_init=True)
```

can you folks take a look?

**Comment by angeloskath** [comment]
So the default behavior is actually exactly PyTorch's see the last comment in the PR. I should have checked whether PT is using Kaiming uniform which they aren't.

Closing the issue but as I mentioned in the PR feel free to open an issue specific to any initialization related problems you may encounter.
